### PR TITLE
Add ValidateAddress RPCClient  to NBitcoin

### PR DESCRIPTION
**Pull request by DanGould**
* tested for bitcoind Mainnet
* Return type for ValidateAddress
* Sister pull to [BreezeProject#25](https://github.com/BreezeHub/BreezeProject/pull/25)

It seems like RPCClientTests.cs has changed (EstimateFeeRate is gone) What's the best process to incorporate this into that lib?

### CODE REVIEW

File: src/NBitcoin.Tests/RPCClientTests.cs [hunk 1]
```diff
@@ -100,6 +100,21 @@ public void CanGetBlockFromRPC()
             }
         }
 
+        [Fact]
+        public void TryValidateAddress()
+        {
+            using (var builder = NodeBuilder.Create())
+            {
+                var node = builder.CreateNode();
```
Comment (noescape00) [inline src/NBitcoin.Tests/RPCClientTests.cs]: 108,111,113 - unvar please.

**Comment by dangershony** [comment]
Apart for @noescape00 comments looks good to me

File: src/NBitcoin/ValidatedAddress.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace NBitcoin
+{
+    public class ValidatedAddress
```
Comment (fassadlr) [inline src/NBitcoin/ValidatedAddress.cs]: I'm not sure whether we should be adding to NBitcoin like this @DanGould . We are in the process of porting over the library to the full node ;)

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs [hunk 3]
```diff
@@ -185,33 +185,33 @@ public BlockHeaderModel GetBlockHeader(string hash, bool isJsonFormat = true)
         /// <returns>JObject containing a boolean indicating address validity</returns>
         [ActionName("validateaddress")]
         [ActionDescription("Returns information about a bech32 or base58 bitcoin address")]
-        public JObject ValidateAddress(string address)
+        public ValidatedAddress ValidateAddress(string address)
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: @DanGould this is already much better. But lets just return JsonResult here. Then, in the method return the boolean like so:

```
var isAddressValid = false;
return JsonResult(isAddressValid);
```

File: src/NBitcoin/RPC/RPCClient.cs [hunk 4]
```diff
@@ -1236,6 +1236,17 @@ public Task SendRawTransactionAsync(byte[] bytes)
 #endregion
 
 #region Utility functions
+        /// <summary>
+        /// Returns information about a base58 or bech32 Bitcoin address
+        /// </summary>
+        /// <param name="address">a Bitcoin Address</param>
+        /// <returns>{ IsValid }</returns>
+        public ValidatedAddress ValidateAddress(BitcoinAddress address)
```
Comment (fassadlr) [inline src/NBitcoin/RPC/RPCClient.cs]: You can really just return a boolean here ;)

File: src/NBitcoin/RPC/RPCClient.cs [hunk 5]
```diff
@@ -1236,6 +1236,17 @@ public Task SendRawTransactionAsync(byte[] bytes)
...
+        {
+            RPCResponse res = SendCommand(RPCOperations.validateaddress, address.ToString());
+            return JsonConvert.DeserializeObject<ValidatedAddress>(res.Result.ToString());
```
Comment (fassadlr) [inline src/NBitcoin/RPC/RPCClient.cs]: And then you don't have to deserialize the result to a class like this. `res.Result` should be a boolean... So just cast it?

**Comment by mikedennis** [comment]
There have been fairly recent changes to NBitcoin with RPC fee rate tests. There now seem to be some estimatesmartfee tests. I don't believe I migrated those across as I think they target a newer version of bitcoin RPC. See https://github.com/MetacoSA/NBitcoin/commit/0d683cef6356ad56d0b5385b7e00d24ee4b56458#diff-280b17ef00043f2fb2b29ad882a40452

File: src/NBitcoin/ValidatedAddress.cs (hunk 2 above)
Comment (bokobza) [reply src/NBitcoin/ValidatedAddress.cs]: I think this is OK, just add `[JsonProperty(PropertyName = "isvalid")]` on top of the property in order to define the exact format of this json field.

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs (hunk 3 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: There is more to it than just a boolean.
Look here:
https://bitcoin.org/en/developer-reference#validateaddress.
For now that's enough, but I think we can easily add more to it.

File: src/NBitcoin/RPC/RPCClient.cs (hunk 4 above)
Comment (bokobza) [reply src/NBitcoin/RPC/RPCClient.cs]: Not really. See my other comment.

**Comment by bokobza** [comment]
@DanGould Are you able to work on this PR? If not, please let us know. Thank you 😃

**Comment by DanGould** [comment]
@bokobza I will have time Thursday & Friday. Classes just started and I've just about found the groove.

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: @bokobza my apologies, I wasn't aware of the bitcoin developer reference! In that case yes returning ValidateAddress is good 👍

File: src/NBitcoin/ValidatedAddress.cs (hunk 2 above)
Comment (fassadlr) [reply src/NBitcoin/ValidatedAddress.cs]: 👍

**Comment by fassadlr** [comment]
@DanGould please check the comments from @noescape00 and once that is fixed we can merge this change 👍

File: src/NBitcoin.Tests/RPCClientTests.cs (hunk 1 above)
Comment (fassadlr) [reply src/NBitcoin.Tests/RPCClientTests.cs]: @DanGould Just fix these and we are good to go 👍

**Comment by bokobza** [comment]
@fassadlr I'd like to see the JsonProperty in ValidatedAddress.cs as well, please. Before you merge this.

**Comment by fassadlr** [comment]
@bokobza will do 💯

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs [hunk 6]
```diff
@@ -205,36 +205,36 @@ public BlockHeaderModel GetBlockHeader(string hash, bool isJsonFormat = true)
...
         {
             if (string.IsNullOrEmpty(address))
                 throw new ArgumentNullException("address");
 
-            var res = new JObject();
-            res["isvalid"] = false;
+            var res = new ValidatedAddress();
+            res.IsValid = false;
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: This is implied if the bool is not specifically set to false ;) so you can remove this line 👍

File: src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs [hunk 7]
```diff
@@ -205,36 +205,36 @@ public BlockHeaderModel GetBlockHeader(string hash, bool isJsonFormat = true)
         /// Returns information about a bitcoin address
         /// </summary>
         /// <param name="address">bech32 or base58 BitcoinAddress to validate.</param>
-        /// <returns>JObject containing a boolean indicating address validity</returns>
+        /// <returns>ValidatedAddress containing a boolean indicating address validity</returns>
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.RPC/Controllers/FullNodeController.cs]: Change this to: 

`/// <returns><see cref="ValidatedAddress"/> containing a boolean indicating address validity.</returns>`

**Comment by bokobza** [comment]
@fassadlr do you mind if we push this as it is? @DanGould doesn't have much time between his classes.
Worst case we can do a quick follow-up PR.

**Comment by fassadlr** [comment]
@bokobza OK sweet 👍

**Pull request merged**